Remove OwnerPaysFee as it's never fully supported

a1q123456 opened the pull request:
## High Level Overview of Change

As instructed, the OwnerPaysFee amendment is never fully supported and therefore we should remove it. However, after some investigation, I found that we can't fully purge everything relating to it because some other features are relying on it. I've removed the feature from [RippleCalc.cpp](https://github.com/XRPLF/rippled/compare/a1q123456/remove-unused-feature-owner-pays-fee?expand=1#diff-141aea7be6c8a950d3785a68ba8b8864e245d4ee287727c9561a503fb44cf009) and set the ownerPaysFee flag to `false` for `flow()`, but this flag is sometimes hardcoded as `true` in some contexts, e.g. XChainBridge.cpp at line 502, CashCheck.cpp at line 448, and CreateOffer.cpp at line 804, which implies we shouldn't fully remove all the business logic relating to it.

I've also removed some unit test cases because they were written for the OwnerPaysFee feature and there are already some test cases for the regular scenario with OwnerPaysFee disabled.

### Context of Change

This change removes unneeded logic and unit tests.

### Type of Change


- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

This change doesn't affect any public protocol.

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

## Test Plan

As this is the first time we have a feature that is both obsolete and unsupported, we should check if any related systems support it. e.g. https://xrpscan.com/amendments and https://js.xrpl.org/

Bronek commented:
@a1q123456 you can fix the formatting errors with 

```
find src include -type f \( -name '*.cpp' -o -name '*.hpp' -o -name '*.h' -o -name '*.ipp' \) -exec clang-format -i {} +
```

You will need clang-format version 1.8.1 in your path. Since you do not need full clang/llvm for formatting only, a simpler way to get it is with `pip`, e.g.

```
pip install clang-format==18.1.8
```

a1q123456 commented:
> @a1q123456 you can fix the formatting errors with
> 
> ```
> find src include -type f \( -name '*.cpp' -o -name '*.hpp' -o -name '*.h' -o -name '*.ipp' \) -exec clang-format -i {} +
> ```
> 
> You will need clang-format version 18.1.8 in your path. Since you do not need full clang/llvm for formatting only, a simpler way to get it is with `pip`, e.g.
> 
> ```
> pip install clang-format==18.1.8
> ```

Thanks Bronek! Will do.

vvysokikh1 commented on src/test/app/Offer_test.cpp:
```diff
@@ -3643,9 +3643,7 @@ class OfferBaseUtil_test : public beast::unit_test::suite
 
         using namespace jtx;
 
-        // The problem was identified when featureOwnerPaysFee was enabled,
-        // so make sure that gets included.
-        Env env{*this, features | featureOwnerPaysFee};
```
I'm not completely sure, but to me this test is relying on this feature completely. Since we remove the feature, this test might be removed as well?

a1q123456 replied:
Double checked those test cases and they are working so we should leave them there in my opinion

vvysokikh1 commented on src/test/app/AMMExtended_test.cpp:
```diff
@@ -1183,9 +1183,7 @@ struct AMMExtended_test : public jtx::AMMTest
 
         using namespace jtx;
 
-        // The problem was identified when featureOwnerPaysFee was enabled,
-        // so make sure that gets included.
-        Env env{*this, features | featureOwnerPaysFee};
```
same comment as in offer_test.cpp

vvysokikh1 reviewed (commented):
General question: are all tests that are being removed failing? If some of them can succeed, they should stay

Bronek reviewed (changes requested):
I spent some time with this PR and I think I see a potential problem here.

The established semantics of "Retired" feature is that it is permanently **enabled** and the code does not support it being disabled. However when we examine how amendment support is handled in `FeatureCollections::registerFeature`, the body of lambda `getAmendmentSupport` makes implied assumption that an `Obsolete` feature is always `Retired` and I think that's incorrect and may potentially cause problems in the future.

I suggest the following change:

```
--- a/src/libxrpl/protocol/Feature.cpp
+++ b/src/libxrpl/protocol/Feature.cpp
@@ -276,7 +276,7 @@ FeatureCollections::registerFeature(
         features.emplace_back(name, f);

         auto const getAmendmentSupport = [=]() {
-            if (vote == VoteBehavior::Obsolete)
+            if (vote == VoteBehavior::Obsolete && support == Supported::yes)
                 return AmendmentSupport::Retired;
             return support == Supported::yes ? AmendmentSupport::Supported
                                              : AmendmentSupport::Unsupported;
```

... and then you will find that you can undo changes in `Feature_test.cpp` and also remove `numObsoleteUnsupportedAmendments` (as they will be now simply considered "unsupported") from `Feature.cpp` and `Feature.h`.

You might also add point 5) in the large comment near the top of `Feature.h` to explain that a feature might be marked `Obsolete`, and that can mean one of two things, depending on `Supported` - either it is in the ledger and it is on its way to become Retired (if it is `Supported::yes`) or the feature is not in the ledger and the code to support it has been removed (if it is `Supported::no`). Feel free to phrase it differently, just note the comment above `retireFeature` in `Feature.cpp`

Bronek commented on include/xrpl/protocol/Feature.h:
```diff
@@ -55,6 +55,10 @@
  *    the scope of these instructions.
  *
+ * 5) A feature marked as Obsolete can mean either:
+ *    1) It is in the ledger (marked as Supported::yes) and it is Retired now
```
they are not "Retired" yet, they are about to become retired. The difference is visible if you scroll to the bottom of `Feature.cpp` file; there are not `feature*` or `fix*` identifiers for a retired feature and there is instead a `[[deprecated]]` `retired*` name, so they cannot be referred to in the code.

a1q123456 replied:
Makes sense, thanks, Bronek!

Bronek commented:
@a1q123456 is this ready ?

vvysokikh1 commented on src/libxrpl/protocol/Feature.cpp:
```diff
@@ -254,7 +254,8 @@ FeatureCollections::registerFeature(
     check(!readOnly, "Attempting to register a feature after startup.");
     check(
-        support == Supported::yes || vote == VoteBehavior::DefaultNo,
+        support == Supported::yes || vote == VoteBehavior::DefaultNo ||
+            vote == VoteBehavior::Obsolete,
         "Invalid feature parameters. Must be supported to be up-voted.");
```
I would like this message to be expanded on. It doesn't describe obsolete case.

vvysokikh1 replied:
ok, I agree on the message. Maybe worth simplifying this conditions to support == Supported::yes || vote != VoteBehavior::DefaultYes

a1q123456 replied:
Fixed

bthomee commented:
@a1q123456 is this ready to merge or are you still working on addressing some comments?

a1q123456 commented:
> @a1q123456 is this ready to merge or are you still working on addressing some comments?

Hey Bart, you're faster than me :). It's ready to merge now.

bthomee commented on include/xrpl/protocol/detail/features.macro:
```diff
@@ -114,7 +114,6 @@ XRPL_FEATURE(DepositAuth,                Supported::yes, VoteBehavior::DefaultYe
 XRPL_FEATURE(Flow,                       Supported::yes, VoteBehavior::DefaultYes)
-XRPL_FEATURE(OwnerPaysFee,               Supported::no,  VoteBehavior::DefaultNo)
 
 // The following amendments are obsolete, but must remain supported
 // because they could potentially get enabled.
```
Should there be a separate section for features that are "obsolete & removed"? This comment suggests it must still remain supported because it potentially can get enabled, but this PR completely removes the code from this feature.

bthomee replied:
Ack - I'll go ahead and merge then.
